### DISCUSSION TITLE
Go routines: database is locked (5) (SQLITE_BUSY)

### DISCUSSION THREAD

**Discussion by seriousm4x** [pocketbase/pocketbase#1795] [Q&A]
Hi,
I'm using pockerbase on a project of mine and [a user reported an error](https://github.com/seriousm4x/UpSnap/issues/45) `database is locked (5) (SQLITE_BUSY)`.
I want to ask if thats an issue with my code or if there's a way to improve my code to avoid database busy errors.

Here is what I do:
1. I iterate over a slice of records (pc's on local network)
2. run 2 go routines on each record:
2.1. ping the pc to see if its up or down
2.2. check pc ports if open or closed

I use go routines because some pings have higher timeouts and can't wait for them. Also i want the frontend to update instantly on changes.

Each go routine contains read and writes to the database. Here is a code snippet:
```go
for _, device := range Devices {
	// ping
	go func(device *models.Record) {
		oldStatus := device.Get("status")
		newStatus := networking.PingDevice(device)
		if newStatus {
			if oldStatus == "offline" || oldStatus == "" {
				device.Set("status", "online")
				app.Dao().SaveRecord(device)
			}
		} else {
			if oldStatus == "online" || oldStatus == "" {
				device.Set("status", "offline")
				app.Dao().SaveRecord(device)
			}
		}
	}(device)

	// scan ports
	go func(device *models.Record) {
		ports, err := app.Dao().FindRecordsByIds("ports", device.GetStringSlice("ports"))
		if err != nil {
			logger.Error.Println(err)
		}
		for _, port := range ports {
			isUp := networking.CheckPort(device.GetString("ip"), port.GetString("number"))
			if isUp != port.GetBool("status") {
				port.Set("status", isUp)
				app.Dao().SaveRecord(port)
				device.RefreshUpdated()
				app.Dao().SaveRecord(device)
			}
		}
	}(device)
}
```
(Reference: https://github.com/seriousm4x/UpSnap/blob/master/backend/cronjobs/cronjobs.go#L38-L72)

Is the error an issue with my code? A problem on pocketbase's side? Is there a better way to implement this?

EDIT: error gets logged in the error check after `ports, err := app.Dao().FindRecordsByIds("ports", device.GetStringSlice("ports"))`

**Comment by ganigeorgiev** [comment]
It is hard to say what is the cause of the issue. Usually the `SQLITE_BUSY` errors are triggered on high concurrent load and/or during mixed read/write transactions. The above doesn't seem to be wrapped in a transaction so it could be something else.

At first glance the above seems ok to me. Some concurrent issues may arise for the [`cronjobs.Devices`](https://github.com/seriousm4x/UpSnap/blob/master/backend/pb/pb.go#L178) because `refreshDeviceList()` is called from various places and you may want to use a mutex, but I doubt that this is the cause for the busy error.

Does it consistently/always trigger the error? Could you provide reproducible steps (_or better minimal reproducible repo_)?

**Reply by seriousm4x** [reply]
Thanks for the input. Good call on `refreshDeviceList()`. I will change it.
I also didn't encounter this personally, but I've asked for more infos [in the issue](https://github.com/seriousm4x/UpSnap/issues/45). I'll update this here if there are more infos.

**Reply by ganigeorgiev** [reply]
I've tried to create a local minimal demo but I wasn't able to reproduce the error executing similar operations as in the above example (aka. firing multiple routines with several read and writes in each). And this makes sense because when there is no transaction it shouldn't trigger the error since both read and writes operates in a non-blocking manner due to the WAL mode.

I've tried to minimize the transaction locking times as much as possible, but still there are some scenarios where a `SQLITE_BUSY` error could occur on heavy concurrent load:

1. If there is read/write during record delete (especially if there is cascade operations over large datasets).

2. If there is read/write during non-admin authenticated record create via the HTTP api. This could happen because we perform a "dry submit" within a mixed transaction with the submitted record data to check the API rules.

We already have auto fail/retry mechanism for the write operations (create/update/delete) but I'll consider in the next v0.13 or v0.14 release to enable it also for the selects (or at least for the record queries).

**Reply by ganigeorgiev** [reply]
@seriousm4x Just to let you know that in the latest [v0.13.0 release](https://github.com/pocketbase/pocketbase/releases/tag/v0.13.0) there is now auto fail/retry for the SELECT queries, which should help keeping the "database is locked" errors to a minimum.

I'm still not able to reproduce the original issue so I'm not sure if the above will fix the reported problem, but if you manage to reproduce it, feel free to provide a minimal repo and I'll further investigate it.

**Reply by seriousm4x** [reply]
Thanks for the info. I will push an update and ask him again. I also wasn't able to reproduce it myself.

**Reply by seriousm4x** [reply]
[Seems to be fixed](https://github.com/seriousm4x/UpSnap/issues/45#issuecomment-1446576544) for him. Thank you!
